### PR TITLE
Support Linear and Bilinear properties for incrementalization in algebra lib

### PR DESCRIPTION
**Pull request by conor-23**

### CODE REVIEW

File: lattices/src/algebra.rs [hunk 1]
```diff
@@ -348,6 +348,59 @@ pub fn idempotency<S: Debug + PartialEq + Clone, const N: usize>(
     Ok(())
 }
 
+/// Defines the linearity property
+/// q is linear with respect to some group operation + if q(a+b) = q(a) + q(b)
+/// This is the same as q being a group homomorphism
+// As defined in the paper "DBSP: Automatic Incremental View Maintenance for Rich Query Languages"
+pub fn linearity<S: Debug + PartialEq + Clone, R: Debug + PartialEq + Clone, const N: usize>(
+    items: &[S; N],
```
Comment (MingweiSamuel) [inline lattices/src/algebra.rs:356]:
```rust
pub fn linearity<S: Debug + PartialEq + Clone, R: Debug + PartialEq + Clone>(
    items: &[S],
```
is simpler and more general - `&[Item; N]` is a reference to a length `N` array, `&[Item]` is a reference to any length array (slice)

File: lattices/src/algebra.rs [hunk 2]
```diff
@@ -348,6 +348,59 @@ pub fn idempotency<S: Debug + PartialEq + Clone, const N: usize>(
...
+// As defined in the paper "DBSP: Automatic Incremental View Maintenance for Rich Query Languages"
+pub fn bilinearity<
+    S: Debug + PartialEq + Clone,
+    R: Debug + PartialEq + Clone,
+    T: Debug + PartialEq + Clone,
+    const N_FOR_F: usize,
+    const N_FOR_SECOND_EXAMPLE: usize,
+>(
+    items_f: &[S; N_FOR_F],
+    items_h: &[T; N_FOR_SECOND_EXAMPLE],
```
Comment (MingweiSamuel) [inline lattices/src/algebra.rs:382]: Similarly for here, you can remove the const lengths

**Review by MingweiSamuel — CHANGES_REQUESTED** [review]

File: lattices/src/algebra.rs (hunk 2 above)
Comment (conor-23) [reply lattices/src/algebra.rs]: fixed

File: lattices/src/algebra.rs (hunk 1 above)
Comment (conor-23) [reply lattices/src/algebra.rs]: fixed

File: lattices/src/algebra.rs [hunk 3]
```diff
@@ -348,6 +348,57 @@ pub fn idempotency<S: Debug + PartialEq + Clone, const N: usize>(
...
 
+/// Defines the linearity property
+/// q is linear with respect to some group operation + if q(a+b) = q(a) + q(b)
+/// This is the same as q being a group homomorphism
+// As defined in the paper "DBSP: Automatic Incremental View Maintenance for Rich Query Languages"
+pub fn linearity<S: Debug + PartialEq + Clone, R: Debug + PartialEq + Clone>(
+    items: &[S],
+    f: impl Fn(S, S) -> S, // The base operation of the algebraic structure for state
+    g: impl Fn(R, R) -> R, // The base operation of the algebraic structure the query q outputs to
+    q: impl Fn(S) -> R, // The query over f that we want to check for linearity (to incrementalize)
```
Comment (MingweiSamuel) [inline lattices/src/algebra.rs:359]: Can we move these comments into the doc comment ```/// ...`

Comment (conor-23) [reply lattices/src/algebra.rs]: resolved in latest commit

**Review by MingweiSamuel — APPROVED** [review]

**Pull request merged**